rcl/include/rcl/event.h

```diff
@@ -0,0 +1,165 @@
+{
+  RCL_SUBSCRIPTION_DEADLINE,
+  RCL_SUBSCRIPTION_LIVELINESS
+} rcl_subscription_event_type_t;
+
+typedef enum rcl_client_event_type_t
+{
+  RCL_CLIENT_EVENT_UNIMPLEMENTED
+} rcl_client_event_type_t;
+
+typedef enum rcl_service_event_type_t
+{
+  RCL_SERVICE_EVENT_UNIMPLEMENTED
+} rcl_service_event_type_t;
+
+
+/// Internal rcl implementation struct.
+struct rcl_event_impl_t;
+
+/// Structure which encapsulates a ROS Subscription.
+typedef struct rcl_event_t
+{
+  struct rcl_event_impl_t * impl;
+} rcl_event_t;
+
+
+/// Return a rcl_event_t struct with members set to `NULL`.
+/**
+ * Should be called to get a null rcl_event_t before passing to
+ * rcl_event_init().
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_event_t
+rcl_get_zero_initialized_event(void);
+
+
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_publisher_event_init(
+  rcl_event_t * event,
+  const rcl_publisher_t * publisher,
+  const rcl_publisher_options_t * options,
+  const rcl_publisher_event_type_t event_type);
+
+
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_subscription_event_init(
+  rcl_event_t * event,
+  const rcl_subscription_t * subscription,
+  const rcl_subscription_options_t * options,
+  const rcl_subscription_event_type_t event_type);
+
+
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_client_event_init(
+  rcl_event_t * event,
+  const rcl_client_t * client,
+  const rcl_client_options_t * options,
+  const rcl_client_event_type_t event_type);
+
+
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_service_event_init(
+  rcl_event_t * event,
+  const rcl_service_t * service,
+  const rcl_service_options_t * options,
+  const rcl_service_event_type_t event_type);
+
+
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_take_event(
+  const rcl_event_t * event,
+  void * event_info);
+
+
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_event_fini(rcl_event_t * event);
+
+
+/// Return the rmw event handle.
+/**
+ * The handle returned is a pointer to the internally held rmw handle.
+ * This function can fail, and therefore return `NULL`, if the:
+ *   - event is `NULL`
+ *   - event is invalid (never called init, called fini, or invalid node)
+ *
+ * The returned handle is made invalid if the event is finalized or if
+ * rcl_shutdown() is called.
+ * The returned handle is not guaranteed to be valid for the life time of the
+ * event as it may be finalized and recreated itself.
+ * Therefore it is recommended to get the handle from the event using
+ * this function each time it is needed and avoid use of the handle
+ * concurrently with functions that might change it.
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | Yes
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[in] event pointer to the rcl event
+ * \return rmw event handle if successful, otherwise `NULL`
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rmw_event_t *
+rcl_event_get_rmw_handle(const rcl_event_t * event);
```

Comment:
Where is this used? Is there a reason we need to expose the rmw layer above the rcl layer?

Reply:
Looks like it is not used anywhere. But neither is `rcl_subscription_get_rmw_handle()`. However, `rcl_publisher_get_rmw_handle()` _is_ used in `rclcpp`...

Let's just leave it here for now.

Reply:
Apparently, this function is needed by the `SET_ADD_RMW` macro in `wait.c`.